pkg/datasource/postgresql.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2024 KubeAGI.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package datasource
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+
+	"github.com/jackc/pgx/v5"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kubeagi/arcadia/api/base/v1alpha1"
+	"github.com/kubeagi/arcadia/pkg/utils"
+)
+
+var (
+	_ Datasource = (*PostgreSQL)(nil)
+)
+
+// PostgreSQL is a wrapper to PostgreSQL
+type PostgreSQL struct {
+	*pgx.Conn
+}
+
+func NewPostgreSQL(ctx context.Context, c client.Client, dc dynamic.Interface, config *v1alpha1.PostgreSQL, endpoint *v1alpha1.Endpoint) (*PostgreSQL, error) {
+	if err := SetPGEnv(ctx, c, dc, config, endpoint); err != nil {
+		return nil, err
+	}
+	conn, err := pgx.Connect(context.Background(), endpoint.URL)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close(ctx)
+	if err := conn.Ping(ctx); err != nil {
+		return nil, err
+	}
+	return &PostgreSQL{conn}, nil
+}
+func (p *PostgreSQL) Stat(ctx context.Context, info any) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *PostgreSQL) Remove(ctx context.Context, info any) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *PostgreSQL) ReadFile(ctx context.Context, info any) (io.ReadCloser, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *PostgreSQL) StatFile(ctx context.Context, info any) (any, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *PostgreSQL) GetTags(ctx context.Context, info any) (map[string]string, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *PostgreSQL) ListObjects(ctx context.Context, source string, info any) (any, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+// SetPGEnv will export all pg setting to environment variable
+func SetPGEnv(ctx context.Context, c client.Client, dc dynamic.Interface, config *v1alpha1.PostgreSQL, endpoint *v1alpha1.Endpoint) error {
```

Comment:
What if we have two postgresql datasources ?  Will they have conflicts in these environment variables?

Reply:
if we have two more postgresql datasources, we should not set env, i guess...

Reply:
well... then we need to find another way to pass the credentials and other configurations to pgx I think

Reply:
Some attempts later, I think it's still a good idea to configure environment variables, delete them after creating the pool, and ensure that multiple configurations don't affect each other when created by using `sync.Mutex` to lock it.

I realized that if I configure the advanced configurations directly, it's equivalent to reimplementing the https://github.com/jackc/pgx/blob/6f2ce92356f35263258ef7baa4cde86a3e1c6255/pgconn/config.go#L229 `ParseConfigWithOptions` function again.